Feature: Full-Duplex Wired Split

petejohanson opened the pull request:
~~Opening this as a draft for now, pending more extensive testing, documentation, etc.~~

~~I've tested this with a very basic ZMK Uno setup, but nothing extensive dailying on a keyboard, etc.~~

Preview docs: https://deploy-preview-2766--zmk.netlify.app/docs/config/split#wired-splits

Has gotten a fair amount of testing at this point, so moving this out of Draft.

<!-- Note: ZMK is generally not accepting PRs for new keyboards. New generic controller PRs *may* still be accepted, please discuss on the Discord server first. -->

## PR check-list

- [x] Branch has a [clean commit history](https://zmk.dev/docs/development/contributing/pull-requests#clean-commit-history)
- [x] Additional tests are included, if changing behaviors/core code that is testable.
- [x] Proper Copyright + License headers added to applicable files (Generally, we stick to "The ZMK Contributors" for copyrights to help avoid churn when files get edited)
- [x] [Pre-commit](https://zmk.dev/docs/development/local-toolchain/pre-commit) used to check formatting of files, commit messages, etc.
- [x] Includes any necessary [documentation changes](https://zmk.dev/docs/development/contributing/documentation).

## Other To-Dos

- [x] Investigate report of stuck/broken taps on peripheral Cirque. Recovered by restarting left side only.
- [x] Address stack size issues for RP2040 exacerbated by this PR?

caksoylar commented:
Related: #1110.

Nick-Munnich reviewed (commented):
Couple of notes on the first commit. Haven't dealt with the split ble code bits enough to be a trustworthy reviewer there, but those bits look fine at a glance.

Not looked at the second commit at all.

petejohanson commented on app/include/zmk/split/transport/types.h:
```diff
@@ -0,0 +1,73 @@
+        struct {
+            uint8_t position;
+            uint8_t pressed;
+        } key_position;
```
I think probably `key_position_event` would maybe be better.

petejohanson replied:
Done.

petejohanson commented on app/include/zmk/split/bluetooth/central.h:
```diff
@@ -1,22 +1,11 @@
-
-#endif // IS_ENABLED(CONFIG_ZMK_SPLIT_PERIPHERAL_HID_INDICATORS)
-
 #if IS_ENABLED(CONFIG_ZMK_SPLIT_BLE_CENTRAL_BATTERY_LEVEL_FETCHING)
 
 int zmk_split_get_peripheral_battery_level(uint8_t source, uint8_t *level);
```
Yeah, that's still on my list to move, just was trying to avoid distraction from the core changes until it settled. Good call out.

petejohanson replied:
Ok, refactored this to cover battery levels over any transport, and added wired handling for this.

petejohanson commented on app/include/zmk/split/central.h:
```diff
@@ -0,0 +1,42 @@
+#if IS_ENABLED(CONFIG_ZMK_SPLIT_BLE)
+
+#include <zmk/ble.h>
+#define BLE_PERIPHERAL_COUNT ZMK_SPLIT_BLE_PERIPHERAL_COUNT
+
+#else
+
+#define BLE_PERIPHERAL_COUNT 0
+
+#endif
+
+#if IS_ENABLED(CONFIG_ZMK_SPLIT_WIRED)
+#define WIRED_PERIPHERAL_COUNT 1
+#else
+#define WIRED_PERIPHERAL_COUNT 0
+#endif
+
+#define ZMK_SPLIT_CENTRAL_PERIPHERAL_COUNT MAX(BLE_PERIPHERAL_COUNT, WIRED_PERIPHERAL_COUNT)
```
This will definitely need changing in the event we do that, but I am not interested in trying to *guess* how that will work, and put something half baked now. I'd rather keep it simple, and refactor once we actually need/use it.

petejohanson commented on app/include/zmk/split/central.h:
```diff
@@ -0,0 +1,42 @@
+#include <zmk/hid_indicators_types.h>
+#endif // IS_ENABLED(CONFIG_ZMK_SPLIT_PERIPHERAL_HID_INDICATORS)
+
+int zmk_split_cental_invoke_behavior(uint8_t source, struct zmk_behavior_binding *binding,
```
What's wrong with `cental`?! /s

(thanks)

Nick-Munnich commented on app/src/split/central.c:
```diff
@@ -0,0 +1,138 @@
+    }
+}
+
+int zmk_split_central_invoke_behavior(uint8_t source, struct zmk_behavior_binding *binding,
```
```suggestion
int zmk_split_central_invoke_behavior(uint8_t target, struct zmk_behavior_binding *binding,
```
Minor suggestion, but I just remembered thinking the other day that this name for this variable (and similar variables in functions around the transport) would be nicer. Feel free to ignore and hit resolve.

petejohanson replied:
I'm on the fence on this one... It requires just a *tad* bit more mental load to understand source/target are the same thing, just from different perspectives, but I can see where you are coming from. Leaving unresolved for now to sleep on it. Thanks.

petejohanson commented on app/include/linker/zmk-split-transport-central.ld:
Ok, so Zephyr offers a functionality called [iterable sections](https://docs.zephyrproject.org/3.5.0/kernel/iterable_sections/index.html) that allow you to define a section, with a certain type expected to be placed in that section, and allows other code to add their own entry(ies) to that section. It's a very convenient way to add extension mechanisms, because you decouple the code that needs to consume a set of items, from the code that needs to include items in the set.

In particular, it allows adding iterable sections to be stored in different places, either RAM or ROM, and handles all the linker magic, so things get placed in the right place in flash or memory, etc. as needed.

So, these files "declare" the iterable sections to the linker, and work along with the other C macros to allow inserting items, iterating the items in a given section at runtime, etc.

petejohanson commented:
Ok, taking this out of draft at this point, since it's gotten a fair amount of use and testing by myself and a few others. I just pushed some of the initial docs work as well, which needs review still.

Nick-Munnich commented on docs/docs/config/split.md:
```diff
@@ -0,0 +1,86 @@
+| Config                                       | Type | Description                                                              | Default |
+| -------------------------------------------- | ---- | ------------------------------------------------------------------------ | ------- |
+| `CONFIG_ZMK_SPLIT`                           | bool | Enable split keyboard support                                            | n       |
+| `CONFIG_ZMK_SPLIT_ROLE_CENTRAL`              | bool | `y` for central device, `n` for peripheral                               |         |
```
```suggestion
| `CONFIG_ZMK_SPLIT_ROLE_CENTRAL`              | bool | `y` for central device, `n` for peripheral                               |  n       |
```

Nick-Munnich commented on docs/docs/config/split.md:
```diff
@@ -0,0 +1,86 @@
+| Config                                | Type   | Description                                                       | Default                                                       |
+| ------------------------------------- | ------ | ----------------------------------------------------------------- | ------------------------------------------------------------- |
+| `CONFIG_ZMK_SPLIT_WIRED`              | bool   | Use wired connection to communicate between split keyboard halves | y (if no BLE split and devicetree is set appropriately)       |
+| `ZMK_SPLIT_WIRED_UART_MODE`           | choice | The choice of which UART mode to use for TX/RX of data.           |                                                               |
```
```suggestion
```
I don't think we expose choice options to people (yet)

Nick-Munnich commented on docs/docs/config/split.md:
```diff
@@ -0,0 +1,86 @@
+| ------------------------------------- | ------ | ----------------------------------------------------------------- | ------------------------------------------------------------- |
+| `CONFIG_ZMK_SPLIT_WIRED`              | bool   | Use wired connection to communicate between split keyboard halves | y (if no BLE split and devicetree is set appropriately)       |
+| `ZMK_SPLIT_WIRED_UART_MODE`           | choice | The choice of which UART mode to use for TX/RX of data.           |                                                               |
+| `ZMK_SPLIT_WIRED_UART_MODE_ASYNC`     | bool   | Async (DMA) mode                                                  | y if the driver supports it (excluding nRF52 with known bugs) |
+| `ZMK_SPLIT_WIRED_UART_MODE_INTERRUPT` | bool   | Interrupt mode                                                    | `y` if the hardware supports it                               |
+| `ZMK_SPLIT_WIRED_UART_MODE_POLLING`   | bool   | Polling mode                                                      | `y` it neither other mode is supported                        |
```
Missing prefix

Nick-Munnich commented on docs/docs/config/split.md:
```diff
@@ -0,0 +1,86 @@
+
+| Config                             | Type | Description                                                 | Default |
+| ---------------------------------- | ---- | ----------------------------------------------------------- | ------- |
+| `ZMK_SPLIT_WIRED_ASYNC_RX_TIMEOUT` | int  | RX Timeout (in microseconds) before reporting received data | 20      |
```
As above

Nick-Munnich commented on docs/docs/config/split.md:
```diff
@@ -0,0 +1,86 @@
+
+| Config                              | Type | Description                                          | Default |
+| ----------------------------------- | ---- | ---------------------------------------------------- | ------- |
+| `ZMK_SPLIT_WIRED_POLLING_RX_PERIOD` | int  | Number of ticks between calls to poll for split data | 10      |
```
As above

Nick-Munnich commented on docs/docs/config/split.md:
```diff
@@ -0,0 +1,86 @@
+
+### Wired Split
+
+Wired splits require a properly configured UART to function. If writing a shield, you may be able to use the standard UART already provided by the board, e.g. `&pro_micro_serial`. See shield overlays section in the [new shield guide](../development/hardware-integration/new-shield.mdx#shield-overlays) for details on the UART node labels provided by various interconnects. If creating your own board ...
```
```suggestion
Wired splits require a properly configured UART to function. If writing a shield, you may be able to use the standard UART already provided by the board, e.g. `&pro_micro_serial`. See the shield overlays section in the [new shield guide](../development/hardware-integration/new-shield.mdx#shield-overlays) for details on the UART node labels provided by various interconnects. If you are creating your own board, or using custom pins for the UART, see the documentation on [pin control](../development/hardware-integration/pinctrl.mdx#additional-examples) to configure the pins for your UART.
```

Nick-Munnich commented on docs/docs/features/split-keyboards.md:
```diff
@@ -6,10 +6,9 @@ sidebar_label: Split Keyboards
+Currently ZMK only supports split keyboards that communicate with each other wirelessly over BLE or with a wired, full-duplex UART.
 
-Supporting split communication over wired protocols is planned, allowing for ZMK split keyboards using non-wireless controllers.
+Future single-wire, half-duplex UART support, which is planned, will allow using wired ZMK with designs like Corne, Sweep, etc. that use only a single GPIO pin for bidirectional communication between split sides.
```
This is the bit that was commented on on Discord. Best to mark it down as experimental initially, and change it later, rather than implying full support from the get-go.

petejohanson replied:
Tweaked the wording here. Thoughts?

petejohanson replied:
Tweaked.

Nick-Munnich reviewed (commented):
Minor comments on the docs.

caksoylar commented on docs/docs/config/split.md:
```diff
@@ -0,0 +1,86 @@
+
+### Wired Split
+
+Wired splits require a properly configured UART to function. If writing a shield, you may be able to use the standard UART already provided by the board, e.g. `&pro_micro_serial`. See shield overlays section in the [new shield guide](../development/hardware-integration/new-shield.mdx#shield-overlays) for details on the UART node labels provided by various interconnects. If creating your own board ...
```
The first one should probably link to https://zmk.dev/docs/development/hardware-integration/pinctrl#predefined-nodes instead, which shows non-GPIO bus labels.

petejohanson replied:
Yeah, good call, forgot that had moved. Updated.

Nick-Munnich commented on docs/docs/config/split.md:
````diff
@@ -0,0 +1,86 @@
+
+### Wired Split
+
+Wired splits require a properly configured UART to function. If writing a shield, you may be able to use the standard UART already provided by the board, e.g. `&pro_micro_serial`. See shield overlays section in the [new shield guide](../development/hardware-integration/new-shield.mdx#shield-overlays) for details on the UART node labels provided by various interconnects. If creating your own board ...
+
+Once you have a properly configured UART device, it needs to be assigned to the `zmk,split-uart` property in the `chosen` node. For example:
+
+```dts
+/ {
+    chosen {
+      zmk,split-uart = &pro_micro_serial;
+    };
+}
+```
````
I think you move this over to the new shield guide, as a subsection under the chosen node (make sure that the headers don't break the TOC on the right) and rephrase it to flow with the new shield guide. Then you can make this text much more succinct.

That way readers of the new shield guide don't have to click to come over here, and then click again to go to pinctrl if they're doing somethign more advanced. Readers of this section should be more prepared to do click-hopping, so to speak.

petejohanson replied:
Ok, so I added to the new shield guide, but didn't really see much way to trim this down without just fully removing it, so I left it as is.

Thoughts?

Nick-Munnich replied:
Resolving this thread so that we can discuss it on the actual file in question,

caksoylar commented on docs/docs/config/split.md:
```diff
@@ -0,0 +1,85 @@
+
+### Wired Split
+
+Wired splits require a properly configured UART to function. If writing a shield, you may be able to use the standard UART already provided by the board, e.g. `&pro_micro_serial`. See [predefined nodes](../development/hardware-integration/pinctrl#predefined-nodes) for details on the UART node labels provided by various interconnects. If you are creating your own board, or using custom pins for th ...
```
```suggestion
Wired splits require a properly configured UART to function. If writing a shield, you may be able to use the standard UART already provided by the board, e.g. `&pro_micro_serial`. See [predefined nodes](../development/hardware-integration/pinctrl.mdx#predefined-nodes) for details on the UART node labels provided by various interconnects. If you are creating your own board, or using custom pins for the UART, see the documentation on [pin control](../development/hardware-integration/pinctrl.mdx#additional-examples) to configure the pins for your UART.
```

Nick-Munnich commented on app/dts/bindings/zmk,wired-split.yaml:
Looking at this file, I'm not a fan of the current method of selecting a UART as chosen. It seems to me that both approaches will exist in the future simultaneously if that is kept. I think while it might be a *little* more complicated *now*, making wired split only go through this node will simplify things in the future. No reason to introduce two ways of doing the same thing from the get-go imo.

petejohanson replied:
Yeah... I'm open to this. I had originally suspected I wouldn't get to anything that required the dedicated node soon enough, but now that I have, I'm open to just documenting the dedicated node and using that as the canonical way.

petejohanson replied:
Pushed this tweak in a few spots, still need to figure out best approach to the new shield guide.

petejohanson replied:
Ok, removed the `chosen` use, tweaked the docs, warned folks on Discord. Now the dedicated node is the only way to configure this.

Nick-Munnich commented on docs/docs/development/hardware-integration/new-shield.mdx:
````diff
@@ -511,6 +511,19 @@ If all of your physical layouts use the same `kscan` node under the hood, you ca
+    };
+};
+```
+
````
Would be good to have a direct link to the predefined nodes here.

petejohanson replied:
Added.

Nick-Munnich reviewed (commented):
Some comments. Still working my way through codey-bits slowly, but can't seem to write two reviews at the same time.

petejohanson commented on app/src/split/wired/wired.h:
```diff
@@ -0,0 +1,95 @@
+                            zmk_split_wired_process_tx_callback_t process_data_cb,
+                            size_t envelope_size);
+
+#endif
+
+#if IS_ENABLED(CONFIG_ZMK_SPLIT_WIRED_UART_MODE_INTERRUPT)
```
I had originally done this, but it makes it much harder to move/reorganize files it they become context dependent like that, at the expense of an extra few `#endif`s. There's also a theoretical time when we have multiple UARTs (e.g. one PIO based and one dedicated PL011 UART on RP2040, and we need to refactor things to allow multiples with the type set in DTS, for example, so I'd prefer to leave this .

Nick-Munnich replied:
I'd suggested this because `ZMK_SPLIT_WIRED_UART_MODE` is a `choice`. I think if there is a world where we would want multiple enabled at the same time, then it should not be marked as `choice`. In that event, we should rather have the Kconfig mark interrupt and async as mutually exclusive flags due to the note on the Zephyr UART page.

Nick-Munnich replied:
If you wanted to ensure that there isn't an unsupported combination using Kconfig, couldn't you just use `$(dt_compat_any_has_prop,<compatible string>,<prop>[,<value>])`? 

I'm always in favor of moving things to be more future-proof while we can do so without breaking user setups (aside from beta testers obv). Especially since it's a *when* and not an *if*.

petejohanson replied:
We can probably nix' the `choice` piece today, and use the fact that the "other modes" all have APIs that should return `-ENOTSUP` when first setting up a device, so we can have multiple enabled and still have fallbacks without duplicating things in DTS.

petejohanson replied:
Fundamentally, I don't think this Kconfig would require any major breakage when we do (eventually, I don't see it happening near term) support multiple UART peripherals, that we should get derailed on this aspect today (as opposed to the `chosen` discussion which I'm leaning toward agreeing with you on)

petejohanson replied:
We could avoid some risk here by avoiding giving the `choice`  a name so no one attempts to override it in board/shield defconfig files.

Nick-Munnich replied:
To be clear, I'm not opposed to the kconfig stuff existing, I just don't think it should exist as a user facing thing. I guess it would make sense to keep the mutually exclusive modes as a choice, but leave polling out of it. Something like
```
config POLL
 default $(dt_compat_any_has_prop,<compatible string>,<prop>[,<value>])

choice UART_FANCY

config ASYNC
 default $(dt_compat_any_has_prop,<compatible string>,<prop>[,<value>])

...
```
and then put all of that in the standard Kconfig, not in Kconfig.defaults.

petejohanson replied:
I don't see what compatible or props you expect to check on? Can you fill that out, because I'm not seeing your vision without it.

Nick-Munnich replied:
$(dt_compat_any_has_prop,zmk,wired-split,uart-mode,polling)
I was imagining something like this. (I do hope I haven't pulled something impossible out of thin air)

petejohanson replied:
Yeah, and my issue with that is that putting that polling mode *in the devicetree* leads to folks who can set a mode in DTS that the UART they're using doesn't support, and we end up with a major mismatch that we probably can only sorta catch at compile time, making it painful at best. The only two ways that "what does my UART support" is surfaced are:

* Each UART driver Kconfig will select some symbols that indicate what is supported, Sadly, this isn't "scoped", but global, so if *any* enabled UART supports async/DMA, then we'll have that Kconfig symbol enabled. See https://github.com/zmkfirmware/zephyr/blob/v3.5.0%2Bzmk-fixes/drivers/serial/Kconfig.sam0#L12 for an example
* At runtime, if you try to use the async API on a serial device that doesn't support async, we'll get back an `-ENOTSUP`.

So... adding some random DTS prop for this then *further* muddies things, which I'd prefer to avoid.

petejohanson replied:
Which is why I suggested we:

* Not over-complicate this now, if we can have something *today* that won't require breakage later,
* Not try to dictate this with additional props, but eventually support having multiple "Mode codes" enabled, that properly fall back to the "less ideal" mode after attempting to use the more efficient mode and checking for `-ENOTSUP` at runtime.

Nick-Munnich replied:
I see, all of that makes sense, thanks for the thorough explanation. I'm still slightly iffy on the mismatch between what the Kconfig states and what the preprocessor states, but you're right in that it could be resolved when it comes to that.

Sorry for the massive sidetracking here

petejohanson replied:
No worries. I *am* going to remove the name/symbol for the choice, so no one attempts to override the default for it, to avoid breakage later once we support multiple here, so still good coming from it!

petejohanson commented on docs/docs/config/split.md:
```diff
@@ -0,0 +1,86 @@
+
+:::note
+
+Hardware UARTs have a few different modes/approaches to sending and receiving data, with different levels of complexity and performance. Not all hardware nor drivers support all modes, so ZMK has code to support different interaction modes with the UART as needed. The default mode should be properly selected based on the platform's report support, but you can choose to override the mode if needed ...
```
Expanded on this a ton, hopefully it's clear.

Nick-Munnich replied:
Perfect. By the sounds of it the nrf52 should use polling until we upgrade to 3.7?

Nick-Munnich replied:
Probably best to have a hazard box in features/split-keyboards, and then link to it everywhere that wired split is mentioned.

petejohanson replied:
Added a warning. Would appreciate your thoughts.

Nick-Munnich commented on docs/docs/features/split-keyboards.md:
```diff
@@ -6,7 +6,7 @@ sidebar_label: Split Keyboards
 
 :::note[Split communication protocols]
-Currently ZMK only supports split keyboards that communicate with each other wirelessly over BLE.
+ZMK supports split keyboards that communicate with each other wirelessly over BLE, and includes experimental wired split support.
```
```suggestion
ZMK supports split keyboards that communicate with each other wirelessly over BLE.
```
Don't think the wired bit is necessary considering the next sentence.

caksoylar commented:
Note for docs: https://zmk.dev/docs/faq#does-zmk-support-wired-split needs to be updated (in this PR or a future one)

petejohanson commented:
> Note for docs: https://zmk.dev/docs/faq#does-zmk-support-wired-split needs to be updated (in this PR or a future one)

Tweaked that note, would appreciate feedback. Thanks!

caksoylar commented on docs/docs/config/split.md:
```diff
@@ -82,12 +82,13 @@ The following settings only apply when using wired split in polling mode:
-    chosen {
-      zmk,split-uart = &pro_micro_serial;
+    wired_split {
+      compatible = "zmk,wired-split";
+      device = <&pro_micro_serial>;
```
```suggestion
        compatible = "zmk,wired-split";
        device = <&pro_micro_serial>;
```

Nick-Munnich replied:
Do we not also mostly have `phandle` types written without the `<>`?

caksoylar replied:
That's the `path` type, usually used for chosen and aliases nodes: https://docs.zephyrproject.org/3.5.0/build/dts/bindings-syntax.html#type

caksoylar commented on docs/docs/faq.md:
```diff
@@ -55,7 +55,7 @@ Please see the [explainer on boards & shields](development/hardware-integration/
 ### Does ZMK support wired split?
 
-Currently, ZMK only supports wireless split, but wired split is possible and we welcome contributions!
+Currently, ZMK only supports wireless split. Experimental wired split support for some specific hardware designs is available for advanced users to test.
```
```suggestion
Currently, ZMK only supports wireless [split keyboards](features/split-keyboards.md). Experimental wired split support for some specific hardware designs is available for advanced users to test.
```

caksoylar commented on docs/docs/features/split-keyboards.md:
```diff
@@ -6,10 +6,17 @@ sidebar_label: Split Keyboards
+
+:::warning[Hot Plugging Cables]
+
+Many popular cables, in particular, TRRS/TRS cables, can cause irreparable damage to controllers if they are inserted or removed when power is already present on them. Whether or not you are using the wired split functionality or not, _never_ insert or remove such a cable when a controller is powered by USB _or_ battery.
```
```suggestion
Many popular wired interconnects used between split keyboards, in particular, TRRS/TRS cables, can cause irreparable damage to controllers if they are inserted or removed when power is already present on them. Whether or not you are using the wired split functionality or not, _never_ insert or remove such a cable when a controller is powered by USB _or_ battery.
```

petejohanson replied:
I'm on the fence using "interconnect" here since we use that term elsewhere to describe board/shield abstractions.

caksoylar replied:
Fair, maybe something like "physical connectors" instead of "wired interconnects"?

Nick-Munnich replied:
Personally I think this particular warning should be made *more* specific, not less, as this problem specifically affects TRS-style cables. Things like USB, RJ45, MIDI, etc are all exempt.

petejohanson replied:
> Personally I think this particular warning should be made _more_ specific, not less, as this problem specifically affects TRS-style cables. Things like USB, RJ45, MIDI, etc are all exempt.

Beyond my original wording?

Nick-Munnich replied:
Ideally yes, but I was thinking about it a bunch yesterday and I couldn't find a phrasing I liked. Essentially making the warning sound more explicit towards the "bad" cable types, rather than warning against all. I think it's fine as-is though.

What I was also considering is a "unless explicitly stated otherwise by the designer" because I think you *can* create a circuit which doesn't have this sort of flaw. I worry that might overcomplicate the warning and invite lots of questions from inexperienced folks though, opening up a can of worms we dont want to tackle (yet)

caksoylar commented on docs/docs/development/hardware-integration/new-shield.mdx:
```diff
@@ -511,6 +511,19 @@ If all of your physical layouts use the same `kscan` node under the hood, you ca
 
 :::
 
+Additionally, if testing the experimental [wired split](../../features/split-keyboards.md) support, you should assign a [predefined](./pinctrl.mdx#predefined-nodes) or [pinctrl configured](./pinctrl.mdx) UART to the `zmk,split-uart` chosen property:
```
Just leaving a comment here as a reminder to update this section.

petejohanson replied:
I added this in a new section, since it's no longer just an additional chosen node... But that means a placeholder text for unibody shields... Thoughts?

caksoylar commented on docs/docs/config/split.md:
```diff
@@ -0,0 +1,94 @@
+
+Following wired [split keyboard](../features/split-keyboards.md) settings are defined in [zmk/app/src/split/wired/Kconfig](https://github.com/zmkfirmware/zmk/blob/main/app/src/split/wired/Kconfig).
+
+:::note
```
I think I'd make these regular paragraphs without the note box, then move "Following wired [split keyboard](../features/split-keyboards.md) settings..." to below, right above the table.

petejohanson replied:
Tweaked. Thought?

caksoylar commented on docs/docs/config/split.md:
```diff
@@ -0,0 +1,94 @@
+| -------------------------------------------- | ---- | ----------------------------------------------------------------- | ------------------------------------------------------------- |
+| `CONFIG_ZMK_SPLIT_WIRED`                     | bool | Use wired connection to communicate between split keyboard halves | y (if no BLE split and devicetree is set appropriately)       |
+| `CONFIG_ZMK_SPLIT_WIRED_UART_MODE_ASYNC`     | bool | Async (DMA) mode                                                  | y if the driver supports it (excluding nRF52 with known bugs) |
+| `CONFIG_ZMK_SPLIT_WIRED_UART_MODE_INTERRUPT` | bool | Interrupt mode                                                    | `y` if the hardware supports it                               |
+| `CONFIG_ZMK_SPLIT_WIRED_UART_MODE_POLLING`   | bool | Polling mode                                                      | `y` it neither other mode is supported                        |
```
```suggestion
| `CONFIG_ZMK_SPLIT_WIRED_UART_MODE_INTERRUPT` | bool | Interrupt mode                                                    | y if the hardware supports it                               |
| `CONFIG_ZMK_SPLIT_WIRED_UART_MODE_POLLING`   | bool | Polling mode                                                      | y it neither other mode is supported                        |
```
Consistency pass

caksoylar commented on docs/docs/config/split.md:
```diff
@@ -0,0 +1,94 @@
+        compatible = "zmk,wired-split";
+        device = <&pro_micro_serial>;
+    };
+}
```
```suggestion
};
```

caksoylar commented on docs/docs/config/split.md:
```diff
@@ -0,0 +1,94 @@
+
+Wired splits require a properly configured UART to function. If writing a shield, you may be able to use the standard UART already provided by the board, e.g. `&pro_micro_serial`. See [predefined nodes](../development/hardware-integration/pinctrl.mdx#predefined-nodes) for details on the UART node labels provided by various interconnects. If you are creating your own board, or using custom pins fo ...
+
+Once you have a properly configured UART device, it needs to be assigned in a new node with a compatible value of `zmk,wired-split`. For example:
```
```suggestion
Once you have a properly configured UART device, it needs to be assigned in a new node with a compatible value of `"zmk,wired-split"`. For example:
```

caksoylar commented on docs/docs/development/hardware-integration/new-shield.mdx:
```diff
@@ -38,6 +38,22 @@ export const SplitInvisTabs = (props) => (
 
 );
 
+export const UnibodyInvisTabs = (props) => (
```
This component doesn't seem necessary and is unused, right?

caksoylar commented on docs/docs/development/hardware-integration/new-shield.mdx:
```diff
@@ -511,6 +527,28 @@ If all of your physical layouts use the same `kscan` node under the hood, you ca
+### Wired Split
+
+<SplitTabs>
+<TabItem value="unibody">This is only required for wired split keyboards</TabItem>
```
```suggestion
<TabItem value="unibody">This is only required for wired split keyboards.</TabItem>
```

caksoylar commented on docs/docs/development/hardware-integration/new-shield.mdx:
```diff
@@ -511,6 +527,28 @@ If all of your physical layouts use the same `kscan` node under the hood, you ca
+<TabItem value="unibody">This is only required for wired split keyboards</TabItem>
+<TabItem value="split">
+
+If testing the experimental [wired split](../../features/split-keyboards.md) support, you should assign a [predefined](./pinctrl.mdx#predefined-nodes) or [pinctrl configured](./pinctrl.mdx) UART to the `device` property of a new node with `compatible` of `zmk,wired-split`:
```
```suggestion
If testing the experimental [wired split](../../features/split-keyboards.md) support, you should assign a [predefined](./pinctrl.mdx#predefined-nodes) or [pinctrl configured](./pinctrl.mdx) UART to the `device` property of a new node with `compatible` value of `"zmk,wired-split"`:
```

caksoylar commented on docs/docs/development/hardware-integration/new-shield.mdx:
```diff
@@ -511,6 +527,28 @@ If all of your physical layouts use the same `kscan` node under the hood, you ca
+      compatible = "zmk,wired-split";
+      device = <&pro_micro_serial>;
+    };
+}
```
```suggestion
};
```

caksoylar commented on docs/docs/development/hardware-integration/new-shield.mdx:
````diff
@@ -511,6 +511,28 @@ If all of your physical layouts use the same `kscan` node under the hood, you ca
+
+```dts
+/ {
+    wired_split {
+      compatible = "zmk,wired-split";
+      device = <&pro_micro_serial>;
+    };
````
```suggestion
    wired_split {
        compatible = "zmk,wired-split";
        device = <&pro_micro_serial>;
    };
```
Sorry, missed this on the previous pass.

caksoylar commented on docs/docs/development/hardware-integration/new-shield.mdx:
````diff
@@ -511,6 +511,28 @@ If all of your physical layouts use the same `kscan` node under the hood, you ca
+```
+
+See the [wired split](../../config/split.md#wired-split) configuration for more details.
+
````
```suggestion

For wireless split keyboards, this step is **not** required and might want to be skipped, especially since the UART pins on your controller might already be in use for other functionality.
```

It doesn't hurt to be clearer here, since people blindly copy paste code blocks and Pro Micro UART pins are D1/D0, I believe.

caksoylar replied:
Open question whether we say "should be skipped" instead.

petejohanson replied:
> Open question whether we say "should be skipped" instead.

I think probably that wording is better, to be really clear.

petejohanson replied:
Added with the tweaked language.

caksoylar reviewed (approved):
One last note, otherwise docs LGTM.

Nick-Munnich commented on app/include/zmk/split/transport/types.h:
```diff
@@ -0,0 +1,78 @@
+    } data;
+} __packed;
+
+enum zmk_split_transport_central_command_type {
+    ZMK_SPLIT_TRANSPORT_CENTRAL_CMD_TYPE_POLL_EVENTS,
+    ZMK_SPLIT_TRANSPORT_CENTRAL_CMD_TYPE_INVOKE_BEHAVIOR,
+    ZMK_SPLIT_TRANSPORT_CENTRAL_CMD_TYPE_SET_PHYSICAL_LAYOUT,
+    ZMK_SPLIT_TRANSPORT_CENTRAL_CMD_TYPE_SET_HID_INDICATORS,
+} __packed;
+
```
I wonder if it might make sense to abstract things in the transport a bit further, taking some ideas from https://github.com/zmkfirmware/zmk/pull/2036#discussion_r1771981768. Probably for a future PR, but might be worth a bit of thought to see if there're any minor changes that could be done to make that work easier in the future.

petejohanson replied:
Yes, I am probably going to do a "custom payload" enhancement eventually, to allow this. We'll probably reserve the lower IDs for future "core" commands/events, and have the upper half of the IDs used for extension. Not for this PR though, as you note.

Nick-Munnich reviewed (commented):
Last pair of minor queries from me. I think it's unlikely I'll notice any further things from looking at it more.

I will say I don't think this should be merged until we've had another release, which I consider #2729 to be a prerequisite for.

paulshir commented on app/src/split/wired/central.c:
```diff
@@ -0,0 +1,335 @@
+#elif IS_ENABLED(CONFIG_ZMK_SPLIT_WIRED_UART_MODE_ASYNC)
+    zmk_split_wired_async_tx(&async_state);
+#else
+    k_work_submit(&wired_central_tx_work);
```
This isn't defined until line 240 so fails building

paulshir commented on app/src/split/wired/central.c:
```diff
@@ -0,0 +1,335 @@
+}
+
+static void read_timer_cb(struct k_timer *_timer) {
+    zmk_split_wired_poll_in(&rx_buf, uart, &publish_events, sizeof(struct event_envelope));
```
Signature has changed. Should it be`zmk_split_wired_poll_in(&rx_buf, uart, &publish_events, NULL, sizeof(struct event_envelope));`?

paulshir commented on app/src/split/wired/peripheral.c:
```diff
@@ -0,0 +1,303 @@
+static K_WORK_DEFINE(send_pending_tx, send_pending_tx_work_cb);
+
+static void wired_peripheral_read_tick_cb(struct k_timer *timer) {
+    zmk_split_wired_poll_in(&chosen_rx_buf, uart, &publish_commands,
```
Signature has changed here. Using the callback instead of the workqueue to get the message expansion.

`zmk_split_wired_poll_in(&chosen_rx_buf, uart, NULL, &process_tx_callback,`

paulshir commented on app/src/split/wired/wired.c:
```diff
@@ -0,0 +1,322 @@
+
+    ring_buf_put_finish(rx_buf, read);
+
+    if (ring_buf_size_get(rx_buf) >= envelope_size) {
```
Envelope size no longer applies here since the update to compress the messages. Should it be`> 0`? Otherwise it takes about 3 messages before any are actually processed.

petejohanson replied:
Yeah, thanks for all the above catches, I'd missed re-testing the polling code after several refactors. About to push some fixes.

paulshir replied:
Thanks, latest fixes work.
Just a small note, you've set this to `>= 0` as opposed to `> 0`.

petejohanson replied:
> Thanks, latest fixes work. Just a small note, you've set this to `>= 0` as opposed to `> 0`.

Indeed, good catch. Fixed.

paulshir reviewed (commented):
Leaving comments here as opposed to discord as it's easier to point to the issues. Tested and debugged the latest changes in polling mode with a rpi pico. Events and commands working with the changes I commented on here.

petejohanson commented on app/include/zmk/split/transport/types.h:
```diff
@@ -0,0 +1,78 @@
+
+    union {
+        struct {
+            uint8_t position;
+            uint8_t pressed;
```
I actually noticed this oddity after the fact as well, but wanted to minimize any *potential* risk to the behavior by also changing this when so much was in flight. Likely this should be unified, and just use the local/central timestamp, but I didn't want to risk any more in this PR.

Nick-Munnich replied:
Personally I'd say it should use the timestamp when the event occurs at the source. I think I'd encourage you to use the same approach between sensor event and key position event either way though; having one transfer the timestamp while the other doesn't feels like it's asking for trouble to appear *somewhere*.

petejohanson replied:
So, the issue here, is if timestamps are used to compare the time between "key presses" somewhere on the central, and a random timestamp that has a different "zero" will *completely* mess with that logic, so there's definitely some good reason to use the central timestamp for when it gets a key press event.

Nick-Munnich replied:
That is a fair argument. Perhaps a TODO comment somewhere, talking about having timestamps taken from source once we have timestamp syncing set up properly? I recall some discussion around that being a necessity for e.g. synced fancy RGB across halves, so we'll probably have that eventually.

petejohanson replied:
> That is a fair argument. Perhaps a TODO comment somewhere, talking about having timestamps taken from source once we have timestamp syncing set up properly? I recall some discussion around that being a necessity for e.g. synced fancy RGB across halves, so we'll probably have that eventually.
> 
> EDIT: I do think that sensor timestamps should be central too though for now, in that case.

Reasonable. Done.

thazhemadam commented on docs/docs/config/split.md:
```diff
@@ -0,0 +1,90 @@
+| `CONFIG_ZMK_SPLIT_WIRED`                     | bool | Use wired connection to communicate between split keyboard halves | y (if no BLE split and devicetree is set appropriately)       |
+| `CONFIG_ZMK_SPLIT_WIRED_UART_MODE_ASYNC`     | bool | Async (DMA) mode                                                  | y if the driver supports it (excluding nRF52 with known bugs) |
+| `CONFIG_ZMK_SPLIT_WIRED_UART_MODE_INTERRUPT` | bool | Interrupt mode                                                    | y if the hardware supports it                                 |
+| `CONFIG_ZMK_SPLIT_WIRED_UART_MODE_POLLING`   | bool | Polling mode                                                      | y it neither other mode is supported                          |
```
```suggestion
| `CONFIG_ZMK_SPLIT_WIRED_UART_MODE_POLLING`   | bool | Polling mode                                                      | y if neither other mode is supported                          |
```

petejohanson replied:
Thanks. Fixed.

FearlessSpiff commented:
This broke my unibody board. Had to fall back to 5ba7e260f49d7277ade882410e7e97fda6e815d4 for it to work again. 
Unfortunatley I can't share my repo as I have other things there I need to fix first. Maybe it's just me?!

nmunnich commented:
> This broke my unibody board. Had to fall back to [5ba7e26](https://github.com/zmkfirmware/zmk/commit/5ba7e260f49d7277ade882410e7e97fda6e815d4) for it to work again. Unfortunatley I can't share my repo as I have other things there I need to fix first. Maybe it's just me?!
> 
> It was not even working over USB. Got some errors in `dmesg` when connecting it.

Could you provide some details such as a repo link so that we can investigate?

FearlessSpiff commented:
I would love to but I pushed some "private" stuff which I first have to get rid of somehow.

petejohanson commented:
> This broke my unibody board. Had to fall back to [5ba7e26](https://github.com/zmkfirmware/zmk/commit/5ba7e260f49d7277ade882410e7e97fda6e815d4) for it to work again. Unfortunatley I can't share my repo as I have other things there I need to fix first. Maybe it's just me?!
> 
> It was not even working over USB. Got some errors in `dmesg` when connecting it.

Can you share any high level details? What controller, etc?

FearlessSpiff commented:
Ok. Fixed it. I am using a Nicenano V2 and a cirque trackpad. But I removed all it's config when I looked for the problem. So here's the repo which works when using a commit before this one: https://github.com/FearlessSpiff/raveniung-zmk-config

Also what gets built with the actions is not working. So that's why I used a build locally.

And here is what dmesg told me:
```
 8511.628105] usb 7-1: device descriptor read/64, error -110
[ 8511.844931] usb 7-1: new full-speed USB device number 8 using xhci_hcd
```

petejohanson commented:
> Ok. Fixed it. I am using a Nicenano V2 and a cirque trackpad. But I removed all it's (trackpad) config when I looked for the problem. So here's the repo which works when using a commit before this one: https://github.com/FearlessSpiff/raveniung-zmk-config

Just to be sure I understand here... Did this work on latest `main` if you remove the cirque bits?

> 
> Also what gets built with the actions is not working. So that's why I used a build locally.
> 
> And here is what dmesg told me:
> ```
>  8511.628105] usb 7-1: device descriptor read/64, error -110
> [ 8511.844931] usb 7-1: new full-speed USB device number 8 using xhci_hcd
> ```


I assume it's also not advertising/connecting over BT?

FearlessSpiff commented:
> > Ok. Fixed it. I am using a Nicenano V2 and a cirque trackpad. But I removed all it's (trackpad) config when I looked for the problem. So here's the repo which works when using a commit before this one: https://github.com/FearlessSpiff/raveniung-zmk-config
> 
> Just to be sure I understand here... Did this work on latest `main` if you remove the cirque bits?
No, it didn't. Sorry to be unclear. I also needed to go back from this change to make it work.
> 
> > Also what gets built with the actions is not working. So that's why I used a build locally.
> > And here is what dmesg told me:
> > ```
> >  8511.628105] usb 7-1: device descriptor read/64, error -110
> > [ 8511.844931] usb 7-1: new full-speed USB device number 8 using xhci_hcd
> > ```
> 
> I assume it's also not advertising/connecting over BT?
Correct. Looked for it but there is no BT advertisement.

petejohanson commented:
@FearlessSpiff Reproduced here, and I just opened https://github.com/zmkfirmware/zmk/pull/2884 to fix this. Will merge as soon as the tests are all green. Sorry for the disruption.

FearlessSpiff commented:
Saw that #2884 got merged and can confirm it is working again for me. Thanks.
